src/sync_finality.rs

```diff
@@ -0,0 +1,341 @@
+use std::sync::{Arc, Mutex};
+
+use anyhow::{anyhow, bail, Context, Result};
+use async_trait::async_trait;
+use avail_subxt::{
+	api::{self, runtime_types::sp_core::crypto::KeyTypeId},
+	avail::{self, Client},
+	primitives::Header,
+	rpc::rpc_params,
+};
+use codec::{Decode, Encode};
+use futures_util::future::join_all;
+use mockall::automock;
+use rocksdb::DB;
+use serde::de::{self};
+use serde::Deserialize;
+use sp_core::{
+	blake2_256,
+	bytes::from_hex,
+	ed25519::{self},
+	twox_128, Pair, H256,
+};
+// use subxt::rpc_params;
+use tokio::sync::mpsc::Sender;
+use tracing::{error, info, trace};
+
+use crate::{
+	data::{get_finality_sync_checkpoint, store_finality_sync_checkpoint},
+	types::{FinalitySyncCheckpoint, GrandpaJustification, SignerMessage, State},
+	utils::filter_auth_set_changes,
+};
+
+#[derive(Debug, Deserialize, Clone)]
+pub struct WrappedJustification(GrandpaJustification);
+
+impl Decode for WrappedJustification {
+	fn decode<I: codec::Input>(input: &mut I) -> std::result::Result<Self, codec::Error> {
+		let j: Vec<u8> = Decode::decode(input)?;
+		let jj: GrandpaJustification = Decode::decode(&mut &j[..])?;
+		Ok(WrappedJustification(jj))
+	}
+}
+
+#[derive(Debug, Deserialize, Decode, Clone)]
+pub struct FinalityProof {
+	/// The hash of block F for which justification is provided.
+	pub block: H256,
+	/// Justification of the block F.
+	pub justification: WrappedJustification,
+	/// The set of headers in the range (B; F] that we believe are unknown to the caller. Ordered.
+	pub unknown_headers: Vec<Header>,
+}
+
+#[derive(Debug, Decode, Clone)]
+pub struct WrappedProof(FinalityProof);
+
+impl<'de> Deserialize<'de> for WrappedProof {
+	fn deserialize<D>(deserializer: D) -> std::result::Result<Self, D::Error>
+	where
+		D: serde::Deserializer<'de>,
+	{
+		let data = from_hex(&String::deserialize(deserializer)?)
+			.map_err(|e| de::Error::custom(format!("{:?}", e)))?;
+		Decode::decode(&mut &data[..]).map_err(|e| de::Error::custom(format!("{:?}", e)))
+	}
+}
+
+#[async_trait]
+#[automock]
```

Comment:
Do we need the mock? If we want to test the flow, we would need to abstract db an client, and this trait would have a bunch of functions.